HW2_1_OOP/Vaulin_HW2_1_OOP.py

```diff
@@ -0,0 +1,362 @@
+
+product = new_arr @ new_arr.T
+
+if (product[0] < 1000).all() and (product[2] > 1000).any():
+    print(product.mean())
+    
+    
+    
+# Наш код здесь
+np = __import__('numpy')
+
+
+matrix = list.__new__(list)
+idx = int.__int__(0)
+while idx.__lt__(100):
+    matrix.__iadd__([list.__call__(range(idx, idx.__add__(10)))])
+    idx = idx.__add__(10)
+    
+selected_columns_indices = list.__call__((filter(lambda x: x in range(1, 5, 2), range(matrix.__len__()))))
+selected_columns = map(lambda x: [x.__getitem__(col) for col in selected_columns_indices], matrix)
+
+arr = np.array(list(selected_columns))
+
+mask = arr.__getitem__((slice(None), 1)).__mod__(3).__eq__(0)
+new_arr = arr.__getitem__(mask)
+
+product = new_arr.__matmul__(new_arr.T)
+
+if product.__getitem__(0).__lt__(1000).__eq__(True).__and__(product.__getitem__(2).__gt__(1000)).__contains__(True).__bool__():
+    print(product.mean().__str__())
+    
+    
+    
+from string import ascii_letters  # чисто чтобы сделать сет всех буковок. нашел много способов от хардкодинга до ord, 
+# но решил не упражняться в изощрениях и импортнул либу, соре
+from abc import ABC, abstractmethod
+
+dna_complement_table = {
+    "a": "t", "A": "T",
+    "t": "a", "T": "A",
+    "g": "c", "G": "C",
+    "c": "g", "C": "G",
+    "n": "n", "N": "N"
+}
+
+rna_complement_table = {
+    "a": "u", "A": "U",
+    "u": "a", "U": "A",
+    "g": "c", "G": "C",
+    "c": "g", "C": "G",
+    "n": "n", "N": "N"
+}
+
+transcription_table = {
+    "a": "a", "A": "A",
+    "t": "u", "T": "U",
+    "u": "t", "U": "T",
+    "g": "g", "G": "G",
+    "c": "c", "C": "C",
+    "n": "n", "N": "N"
+}
+
+
+# Наш код здесь
+class BiologicalSequence(ABC):
+    @abstractmethod
+    def __len__(self):
+        pass
+
+    @abstractmethod
+    def __getitem__(self):
+        pass
+
+    @abstractmethod
+    def __repr__(self):
+        pass
+
+    @abstractmethod
+    def __str__(self):
+        pass
+
+    @abstractmethod
+    def check_alphabet(self):
+        pass
+
+
+class CentralDogmaSequence(BiologicalSequence):
+    def __init__(self, seq, alphabet=ascii_letters):
+        self.seq = seq
+        self.alphabet = set(alphabet)
+
+    def __len__(self):
+        return len(self.seq)
+
+    def __getitem__(self, idx):
+        return self.seq.__getitem__(idx)
+
+    @abstractmethod
+    def __repr__(self):
+        pass
+
+    def __str__(self):
+        return str(self.seq)
+
+    def check_alphabet(self):
+        return set(self.seq) <= self.alphabet
+
+
+class NucleicAcidSequence(CentralDogmaSequence):
+    def __init__(self, seq, alphabet='ATGCU', complement_table=dna_complement_table):
+        super().__init__(seq, alphabet)
+        self.complement_table = complement_table
+
+    def __repr__(self):
+        return f"5'-{self.seq}-3'"
+
+    def get_complement(self):
+        if self.check_alphabet():
+            return NucleicAcidSequence(''.join([self.complement_table[base] for base in self.seq]))

```

Comment:
Решение описанной в комменте проблемы
```suggestion
            return type(self)(''.join([self.complement_table[base] for base in self.seq]))

```

Reply:
Вызвали от РНК? Будет РНК. Вызвали от ДНК? Будет ДНК. Вызвали от чего-то ещё? Будет что-то ещё